### PR TITLE
hyprpaper: init at unstable-2022-07-04

### PR DESCRIPTION
**Pull request by wozeparrot**
###### Description of changes

Adds [hyprpaper](https://github.com/hyprwm/hyprpaper), a blazing fast wayland wallpaper utility. This is supposed to go alongside Hyprland but will work with any wlroots compositor.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by viraptor — COMMENTED** [review]
Result of `nixpkgs-review pr 180192` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>hyprpaper</li>
  </ul>
</details>

Minor, but when running, the app prints out where it was built from: https://github.com/hyprwm/hyprpaper/blob/ae2c232afcbac705a1621c38be499ea302b0628f/src/main.cpp#L6
The current build process does not pass those variables in: https://github.com/hyprwm/hyprpaper/blob/fdaea3d78af54b18c70e04f691bb035d02e5ab6f/CMakeLists.txt#L21

Would be nice to get `rev` in there.

**Comment by viraptor** [comment]
As far as I can tell the `execute_process` is not conditional, right? As in, you can't just set the variable directly without modifying the cmake file? If that's the case, then probably allowing git to work would be better than a fragile rewriting of the script.

**Comment by ShadowRZ** [comment]
Stanisław Pitucha writes:

> As far as I can tell the `execute_process` is not conditional, right?
> As in, you can't just set the variable directly without modifying the
> cmake file? If that's the case, then probably allowing git to work
> would be better than a fragile rewriting of the script.

Yes. I tested `execute_process` will clobber the variable, so it should
make sense to allow Git to work.

File: pkgs/tools/wayland/hyprpaper/default.nix [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    repo = pname;
+    rev = "e15912e9817d79bb988085c88e313fac5ab60940";
+    sha256 = "sha256-UZSRcj+CckUDllBtmlIcwA+xXUonpJZl3zC151IV3f0=";
+  };
+
+  nativeBuildInputs = [
+    cmake
+    pkg-config
+    wayland-scanner
+  ];
```
Comment (ShadowRZ) [inline pkgs/tools/wayland/hyprpaper/default.nix:30]:
```suggestion
, wlroots
, gitMinimal
}:

stdenv.mkDerivation rec {
  pname = "hyprpaper";
  version = "unstable-2022-07-04";

  src = fetchFromGitHub {
    owner = "hyprwm";
    repo = pname;
    rev = "e15912e9817d79bb988085c88e313fac5ab60940";
    leaveDotGit = true;
    sha256 = "sha256-1P4U6fSaiiUYAvoL8EdFPBbtLh5v/S2RMSuSpbISGFc=";
  };

  nativeBuildInputs = [
    cmake
    pkg-config
    wayland-scanner
    gitMinimal
  ];
```

With this change `hyprpaper` can know which commit it was built.
(Yes I calculated the hash for you, you can always re-calculate the hash again by removing the `sha256` line.)

**Review by ShadowRZ — CHANGES_REQUESTED** [review]

File: pkgs/tools/wayland/hyprpaper/default.nix [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+stdenv.mkDerivation rec {
+  pname = "hyprpaper";
+  version = "unstable-2022-07-04";
+
+  src = fetchFromGitHub {
+    owner = "hyprwm";
+    repo = pname;
+    rev = "e15912e9817d79bb988085c88e313fac5ab60940";
+    leaveDotGit = true;
```
Comment (ckiee) [inline pkgs/tools/wayland/hyprpaper/default.nix:23]: This is non-deterministic. Apparently. Sometimes. 

If there's a simple `substituteInPlace` you could add to hardcode the `version` to the Nix one that'd be nice.

File: pkgs/tools/wayland/hyprpaper/default.nix [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  '';
+
+  installPhase = ''
+    mkdir -p $out/bin
+    install -m755 ./hyprpaper $out/bin
+  '';
+
+  meta = with lib; {
+    homepage = "https://github.com/hyprwm/hyprpaper";
+    description = "A blazing fast wayland wallpaper utility.";
```
Comment (ckiee) [inline pkgs/tools/wayland/hyprpaper/default.nix:53]: https://github.com/NixOS/nixpkgs/blob/74499b178ebfbec52639f4aae4b5543df7bf27b2/CONTRIBUTING.md#submitting-changes
```suggestion
    description = "A blazing fast wayland wallpaper utility";
```

**Review by ckiee — CHANGES_REQUESTED** [review]

File: pkgs/tools/wayland/hyprpaper/default.nix (hunk 2 above)
Comment (wozeparrot) [reply pkgs/tools/wayland/hyprpaper/default.nix]: I'm only patching in the commit hash as I don't think there is a way to get the commit message without `.git`.

**Review by ckiee — APPROVED** [review]

**Comment by ckiee** [comment]
Squash-merged since commit history is kept clean around here <3

**Pull request merged**